state/kvcache.go

```diff
@@ -0,0 +1,111 @@
+package state
+
+import (
+	"container/list"
+	"fmt"
+
+	cmn "github.com/tendermint/tmlibs/common"
+)
+
+// KVCache is a cache that enforces deterministic sync order.
+type KVCache struct {
+	store    KVStore
+	cache    map[string]kvCacheValue
+	keys     *list.List
+	logging  bool
+	logLines []string
+}
+
+type kvCacheValue struct {
+	v []byte        // The value of some key
+	e *list.Element // The KVCache.keys element
+}
+
+// NOTE: If store is nil, creates a new MemKVStore
```

Comment:
// NewKVCache creates a new KVCache from store.
// If store is nil, it creates a new MemKVStore.